suite/tests/client-interface/mangle_suspend.dll.c

```diff
@@ -0,0 +1,190 @@
+    ptr_int_t val1, val2;
+    *user_data = NULL;
+
+    if (translating)
+        return DR_EMIT_DEFAULT;
+
+#ifdef X86_64
+    /* Look for duplicate mov immediates telling us which subaction we're in */
+    for (inst = instrlist_first_app(bb); inst != NULL; inst = instr_get_next_app(inst)) {
+        if (instr_is_mov_constant(inst, prev_was_mov_const ? &val2 : &val1)) {
+            if (prev_was_mov_const && val1 == val2 &&
+                val1 != 0 && /* rule out xor w/ self */
+                opnd_is_reg(instr_get_dst(inst, 0)) &&
+                opnd_get_reg(instr_get_dst(inst, 0)) == SUSPEND_TEST_REG) {
+                *user_data = (void *)val1;
+                instrlist_meta_postinsert(bb, inst, INSTR_CREATE_label(drcontext));
+            } else
+                prev_was_mov_const = true;
+        } else
+            prev_was_mov_const = false;
+    }
+#else
+        /* XXX i#3307: port to ARM if possible. */
+#endif
+    return DR_EMIT_DEFAULT;
+}
+
+static void
+suspend_func()
+{
+#ifdef X86_64
+    void **drcontexts = NULL;
+    uint num_suspended;
+    if (dr_suspend_all_other_threads(&drcontexts, &num_suspended, NULL)) {
+        if (num_suspended != 1)
+            CHECK(false, "num_suspended unexpected!");
+        dr_mcontext_t mc = { sizeof(mc), DR_MC_INTEGER | DR_MC_CONTROL };
+        if (!dr_get_mcontext(drcontexts[0], &mc))
+            CHECK(false, "dr_get_mcontext failed!");
+        /* Check xip such that we only perform the check if we are in the loop
+         * body of the test, in order to avoid races.
+         */
+        if (mc.xip >= add_instr_pc - 7 && mc.xip <= add_instr_pc + 24) {
+            if (add_instr_pc == mc.xip) {
+                if (LOOP_COUNT_REG_MC != 1) {
+                    CHECK(false, "loop count reg expected to be 1");
+                    exit(1);
+                }
+            } else if (LOOP_COUNT_REG_MC != 2) {
+                CHECK(false, "loop count reg expected to be 2");
+                exit(1);
+            }
+        }
+        int resume_count = 0;
+        while (!dr_resume_all_other_threads(drcontexts, num_suspended)) {
+            if (resume_count++ == MAX_RESUME_COUNT) {
+                CHECK(false, "resume failed!");
+                exit(1);
+            }
+        }
+    }
+#else
+    /* XXX i#3307: port to ARM if possible. */
+#endif
+}
+
+static dr_emit_flags_t
+event_app_instruction(void *drcontext, void *tag, instrlist_t *bb, instr_t *inst,
+                      bool for_trace, bool translating, void *user_data)
+{
+    ptr_int_t subaction = (ptr_int_t)user_data;
+
+    if (translating)
+        return DR_EMIT_DEFAULT;
+
+    if (subaction == TEST_VAL_C) {
+        if (instr_is_label(inst)) {
+#ifdef X86_64
+            byte *pc = tag;
+            instr_t *instr = instr_create(drcontext);
+            while (true) {
+                instr_reset(drcontext, instr);
+                byte *prev_pc = pc;
+                pc = decode(drcontext, pc, instr);
```

Comment:
This is unusual: DR has already decoded the block.  Normally you would walk the instrlist.  Is there a reason not to walk it?

Reply:
Wasn't clear to me. Fixed!